commit 458eb4d225a47caf3671eaa7b484bd9e67354d89
Author: hsiao-ming.chang <hsiao-ming.chang@wisecore.com.tw>
Date:   Tue Mar 3 10:56:34 2020 +0800

    Add building configuration of openssl.
    Get tag of CCM for decryption.

diff --git a/lib/openssl-1.1.1d/BUILD.sh b/lib/openssl-1.1.1d/BUILD.sh
new file mode 100755
index 0000000..3f6ebde
--- /dev/null
+++ b/lib/openssl-1.1.1d/BUILD.sh
@@ -0,0 +1,20 @@
+#!/bin/sh
+set -x
+./Configure --config=custom.conf  \
+	--cross-compile-prefix=aarch64-linux-gnu- \
+    custom-linux-aarch64 \
+	no-dso \
+	no-hw \
+	no-engine \
+	no-dynamic-engine \
+	no-tests \
+	no-threads \
+	no-shared \
+	no-tests \
+	no-sock \
+	no-ssl2 \
+	no-ssl3 \
+	no-comp \
+	no-idea \
+	no-afalgeng \
+	no-asan \
diff --git a/lib/openssl-1.1.1d/crypto/evp/e_aes.c b/lib/openssl-1.1.1d/crypto/evp/e_aes.c
index 1db346f..6ac918b 100644
--- a/lib/openssl-1.1.1d/crypto/evp/e_aes.c
+++ b/lib/openssl-1.1.1d/crypto/evp/e_aes.c
@@ -3576,7 +3576,7 @@ static int aes_ccm_ctrl(EVP_CIPHER_CTX *c, int type, int arg, void *ptr)
         return 1;
 
     case EVP_CTRL_AEAD_GET_TAG:
-        if (!EVP_CIPHER_CTX_encrypting(c) || !cctx->tag_set)
+        if (!cctx->tag_set)
             return 0;
         if (!CRYPTO_ccm128_tag(&cctx->ccm, ptr, (size_t)arg))
             return 0;
@@ -3730,9 +3730,11 @@ static int aes_ccm_cipher(EVP_CIPHER_CTX *ctx, unsigned char *out,
         return len;
     }
 
+#if 0
     /* The tag must be set before actually decrypting data */
     if (!EVP_CIPHER_CTX_encrypting(ctx) && !cctx->tag_set)
         return -1;
+#endif
 
     /* If not set length yet do it */
     if (!cctx->len_set) {
@@ -3753,17 +3755,23 @@ static int aes_ccm_cipher(EVP_CIPHER_CTX *ctx, unsigned char *out,
         if (cctx->str ? !CRYPTO_ccm128_decrypt_ccm64(ccm, in, out, len,
                                                      cctx->str) :
             !CRYPTO_ccm128_decrypt(ccm, in, out, len)) {
-            unsigned char tag[16];
-            if (CRYPTO_ccm128_tag(ccm, tag, cctx->M)) {
-                if (!CRYPTO_memcmp(tag, EVP_CIPHER_CTX_buf_noconst(ctx),
-                                   cctx->M))
-                    rv = len;
+            if (cctx->tag_set) {
+                unsigned char tag[16];
+                if (CRYPTO_ccm128_tag(ccm, tag, cctx->M)) {
+                    if (!CRYPTO_memcmp(tag, EVP_CIPHER_CTX_buf_noconst(ctx),
+                                    cctx->M))
+                        rv = len;
+                }
+                cctx->tag_set = 0;
+            } else {
+                rv = len;
             }
         }
         if (rv == -1)
             OPENSSL_cleanse(out, len);
         cctx->iv_set = 0;
-        cctx->tag_set = 0;
+#if 0
+#endif
         cctx->len_set = 0;
         return rv;
     }
diff --git a/lib/openssl-1.1.1d/custom.conf b/lib/openssl-1.1.1d/custom.conf
new file mode 100644
index 0000000..54e0b4d
--- /dev/null
+++ b/lib/openssl-1.1.1d/custom.conf
@@ -0,0 +1,6 @@
+(
+    'custom-linux-aarch64' => {
+        inherit_from    => [ 'linux-aarch64' ],
+        dso_scheme     => undef,
+    }
+);
\ No newline at end of file
